### PR TITLE
Implemented visit_AsyncFunctionDef in the AST->ASR visitor

### PR DESCRIPTION
**Pull request by ujjwaltwitx**
Solves issue no #2312. Implemented visit_AsyncFunctionDef function both in the SymbolTableVisitor and BodyVisitor class.

### CODE REVIEW

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 1]
```diff
@@ -4051,6 +4051,17 @@ class SymbolTableVisitor : public CommonVisitor<SymbolTableVisitor> {
...
 
+    void visit_AsyncFunctionDef(const AST::AsyncFunctionDef_t &/*x*/){
+        try
+        {
+            // to be implemented
+        }
+        catch(const std::exception& e)
+        {
+            std::cerr << e.what() << '\n';
+        }
```
Comment (certik) [inline src/lpython/semantics/python_ast_to_asr.cpp:4062]: You have to report the error like this:
```suggestion
            throw SemanticError("The `async` keyword is currently not supported", x.base.base.loc); 
```

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 2]
```diff
@@ -4823,6 +4834,19 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
...
+        try
+        {
+            // BodyVisitor for visit_AsyncFunctionDef to be implemented
+        }
+        catch(const std::exception& e)
+        {
+            std::cerr << e.what() << '\n';
+        }
+        
+    }
```
Comment (certik) [inline src/lpython/semantics/python_ast_to_asr.cpp:4847]: Instead of declaring it twice, declare it just once in the base class.

**Comment by certik** [comment]
Let's also add a test for this in `tests/errors`.

File: src/lpython/semantics/python_ast_to_asr.cpp (hunk 1 above)
Comment (ujjwaltwitx) [reply src/lpython/semantics/python_ast_to_asr.cpp:4062]: Thanks @certik. I will make the necessary changes

**Comment by ujjwaltwitx** [comment]
> Let's also add a test for this in `tests/errors`.

@certik , can you please guide me a bit on how to add the test for the async implementation in tests/errors.  I have created the file `test_async.py` and it contains the below given code
```
async def test_async():
    print("done")

test_async()
```

But when I run ./run_tests.py, the file doesn't get executed.

**Comment by certik** [comment]
You have to add it to `tests/tests.toml`

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 3]
```diff
@@ -4051,6 +4051,18 @@ class SymbolTableVisitor : public CommonVisitor<SymbolTableVisitor> {
...
 
+    void visit_AsyncFunctionDef(const AST::AsyncFunctionDef_t &x){
+        try
+        {
+            // to be implemented
+        }
+        catch(const std::exception& e)
+        {
+            std::cerr << e.what() << '\n';
+        }
```
Comment (certik) [inline src/lpython/semantics/python_ast_to_asr.cpp:4062]: Remove these.

File: src/lpython/semantics/python_ast_to_asr.cpp (hunk 2 above)
Comment (certik) [reply src/lpython/semantics/python_ast_to_asr.cpp:4848]: Remove this function completely.

**Comment by certik** [comment]
Very good, it's almost done. Left couple more comments.

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 4]
```diff
@@ -4051,6 +4056,8 @@ class SymbolTableVisitor : public CommonVisitor<SymbolTableVisitor> {
     // Implement visit_Global for Symbol Table visitor.
     void visit_Global(const AST::Global_t &/*x*/) {}
 
+    
+
```
Comment (certik) [inline src/lpython/semantics/python_ast_to_asr.cpp:4060]: Remove white space.

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 5]
```diff
@@ -4823,6 +4830,7 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
         tmp = nullptr;
     }
 
+
```
Comment (certik) [inline src/lpython/semantics/python_ast_to_asr.cpp:4833]: remove white space

**Comment by certik** [comment]
It looks good! Last two issues, see my comments above.

**Review by certik — APPROVED** [review]
This looks great now, thank you!

**Pull request merged**